[DOC-3637] New delegate troubleshooting section, new troubleshooting certificate issues topic (#3091)
harness/developer-hub: docs/platform/2_Delegates/troubleshooting/certificate-issues.md

---

> title: Certificate issues

Reviewer: ```suggestion
title: Delegate certificate issues
```

Reviewer: just adding this so it's clearer in search results

description: Troubleshoot common delegate certificate issues.
# sidebar_position: 1
---

This topic provides solutions for common delegate certificate issues.

## Delegate fails to register

> In some scenarios, the delegate may start but fail to register. There are two common exceptions that might occur an `SSLHandshakeException` and a signature check failure.

Reviewer: ```suggestion
In some scenarios, the delegate might start to register and then fail. There are two common exceptions that might occur: an `SSLHandshakeException` and a signature check failure.
```

### Handshake exception

You might experience a `javax.net.ssl.SSLHandshakeException: unable to find valid certification path to requested target` exception.

This typically means the Java truststore file doesn’t have the required certificate to connect to Harness Manager because of a missing Certificate Authority (CA).

#### Handshake exception solutions

To resolve the handshake exception, do the following:

1. Run to the command below to test the certificate chain you used to install Harness Manager.

   ```
   curl -cacerts path/to/ca-certs/file https://<MANAGER_HOST>/api/account/<ACCOUNT_ID>/status
   ```

> 2. Install the certificate chain to the delegate.

Reviewer: ```suggestion
2. Install the certificate on the delegate.
```

Reviewer: the "chain" is from the cert to the CA that created it, and then to the trusted root CA. You don't install a chain. The cert contains the chain links in its definition.

   1. If your delegate version is 803xx or later, mount the custom certificates to a path in the container and launch the delegate with the `CUSTOM_CERTS_PATH` environment variable to the path inside the container.

   2. If your delegate version is earlier than 803xx, go to [Install delegates with custom certificates](/docs/platform/delegates/secure-delegates/install-delegates-with-custom-certs/). Remember to add the certificates to both the Java truststore and the pod.

3. Follow the appropriate steps below, based on whether you use the OpenSSL tool.

> **Use the OpenSSL tool**

Reviewer: since this isnt' a heading and the section is very long, consider putting it and the subsequent "When the OpenSSL tool isn't present" sections in folds. Just a thought.

   2. If your delegate version is earlier than 803xx, go to [Install delegates with custom certificates](/docs/platform/delegates/secure-delegates/install-delegates-with-custom-certs/). Remember to add the certificates to both the Java truststore and the pod.

**When the OpenSSL tool isn't present**

To resolve the exception when OpenSSL tool isn't present, do the following:

1. Try to install OpenSSL.

   1. Exec into the delegate.
   2. Run the following.
      
      ```
      microdnf install openssl
      ```

       Depending on your environment, OpenSSL installation may not succeed.
    
    3. If the installation succeeds, following the OpenSSL steps above. If the installation fails, continue with the steps below.

2. Use the cURL commands below to find the Common Name (CN) servers that are missing in your CA bundle.
   
   ```
   curl -vk <YOUR_URL>
   ```

   ```
   curl -vk <YOUR_URL>
   ```

3. Find the certificate for each CN by going to the domain in your browser or download the certificate.

### Signature check failure

In some scenarios, you might experience a `signature check failed: Signature length not correct: got 512 but was expecting 256` exception.

> This exception occurs because the length of the public key is not the same as the length of the signature. During the TLS handshake, the signature received by the delegate (client side) is the certificate sent by the server. The public key is from the truststore file where the delegate loads during startup. The delegate is not installed with the correct CA certificates that match the server side.

Reviewer: ```suggestion
This exception occurs because the length of the public key is not the same as the length of the signature. During the TLS handshake, the signature received by the delegate (client side) is the certificate sent by the server. The public key is from the truststore file where the delegate loads during startup. The issue can occur when the delegate is not installed with CA certificates that match the server side correctly.
```
